### ISSUE TITLE
First battle balancing problem

### ISSUE THREAD

**Issue by rjaguar3** [grrk-bzzt/kq-fork#31]
The player's first battle appears to be extremely difficult and unbalanced, and the player's probability of success is highly dependent on random chance in picking a few weak enemies for the first battle.  Because Noslom's hit points are so low that he cannot afford to take two rounds of enemy attacks and survive, he is basically forced to cast Cure1 every turn and hope for a miss (or misses, in the case of multiple enemies) allowing him to actually damage the enemies.

**Comment by peterhull90** [comment]
Interesting point; do you mean the first battle in the wooded area or once you go down the tunnel? It's been a really long time since I started a game of KQ from scratch but IIRC you need to buy stuff in  Randen then complete a few fights on the world map before tackling the tunnel.

**Comment by rjaguar3** [comment]
I mean in the wooded area just after leaving town (which is where I went when I began the game and I suspect most other players did too).  I noticed it to be difficult myself, and, from looking at the few videos of KQ on YouTube, I'm apparently not the only one who had trouble:
* https://www.youtube.com/watch?v=SjdIi2KQAEY&t=213s
* https://www.youtube.com/watch?v=_II8FymOETI&t=377s
* https://www.youtube.com/watch?v=QW7-I1Ak6Gg&t=532s

**Comment by OnlineCop** [comment]
Once you leave the initial manor, no monsters should attack you till you get to the town. Once you do, the "strong-scented oils" wear off (a flag is set to allow monsters to attack when you leave town again). If you leave town at that point, monsters can attack you anywhere on the world map.

The intent is that a player will enter the town, upgrade weapons/armor with the money given them by Hunert (the butler), and then head back out and fight/level up after they are better-equipped. If careful, they should be able to heal themselves either from their own magic OR from the inn or Medicinal Herbs they should be able to afford after these battles.

Monster battles are determined in an "encounters" table that defines how many monsters to use, as well as which monsters will be in the line-up. Some zones have multiple entries in this table, so it randomly chooses one of the possible battle scenarios. There is currently no logic that considers the player's level when determining which monsters attack, so we can't make fights easier for low-level players and harder for high-level ones.

In that particular area of the map, the monsters that are assigned to that zone are randomly chosen between 1-2 orcs or 1-2 goblins, or a combination thereof. To make it easier to battle, we could strip away the encounters that define 2 or more enemies in it, so the player ONLY encounters a single orc or a single goblin.

When a player returns much later in the game to the same zone, they will still encounter the same battles, so we may have to consider that.

**Comment by peterhull90** [comment]
Hmm that is pretty brutal. I remember this was a bit of a  problem with KQ - the battle difficulty was like a cliff edge - either very difficult or, once you'd levelled up, very easy. Or possibly it's just that I was never very good at it.

We could make the first entry in the encounter table a bit gentler, or give Noslom a bit more HP and/or cash at the start.

**Comment by OnlineCop** [comment]
There are a few "hidden" caches of money that can be found in the first town, as well as a Mace (weapon) in the grotto north of the town and manor. Maybe those could be made more prominent so new players can get them right off the bat. In town, there are two paths at the top-right corner of the map that dead-end with some flowers. Searching those flowers rewards the player with extra money. If changing the battles isn't feasible, maybe changing those flowers to easy-to-see chests would work. In the grotto, above and to the right of the fire, there's a hidden path through the trees that leads to the mace. Maybe that could be made into a chest as well, and the tree "overhangs" could be trimmed enough so the path is more visible.

When I play, I usually grab those things, equip the best weapons, and sell off the rest of my supply. You can even find a free spell book that can be learned or sold for some more immediate cash to use for Medicinal Herbs or stays at the Inn. That gives enough gold to buy some decent armor so even a Level 1 player can withstand hits from 3 monsters at once.

I haven't really worked much with the battle system, but I could probably figure out how to nerf the battle around town to only include 1 enemy per encounter. We could also just start off the player with more gold from Hunert (a 1-line fix to the manor.lua file) of better starting gear (that may be hard-coded, but it may also be set up in a .lua file so no recompile would be necessary).

Thoughts?

**Comment by peterhull90** [comment]
I think it's great that there's a better way to do it, if the player takes the time to explore the town properly. I don't think we should change the map to reveal those things

The problem I can see is that someone might give up on the game if they get dinged twice or three times in a row in the first combat. I have the feeling that games these days (and I know I am sounding like an old man) tend to have a much gentler start than they used to. Remember KQ is 15 years old!

To get rid of the encounters with 2 enemies is easy, just edit `encount.csv` and recompile. The fields idx1, idx2, ... represent the monsters so if you want only one, make sure all the fields apart from idx1 are zero.

Other solutions not considered are some kind of tweak to make sure you can't be killed in the first couple of battles (I am thinking something to bias the random numbers in your favour) or maybe a hint system if you do get killed early on (e.g. tell them to search the first town). They could get a hint and instant resurrection so they don't have to go through the manor scene again.

**Comment by OnlineCop** [comment]
Looking at the code, the file `other/encount.csv` is never actually used or referenced. The data is duplicated in code in [src/res.c](https://github.com/grrk-bzzt/kq-fork/blob/master/src/res.c#L1404) in the `s_erow erows[NUM_ETROWS]` table. I believe the only reason the CSV file exists was so they could look at and adjust the values in a spreadsheet before exporting the data (probably manually or semi-manually) back to a C object.

**Comment by peterhull90** [comment]
Right. There's a script called `kq2h.pl` which converted the csv to a C fragment, then there was a stanza in `makefile.all` (now deleted) which converted those fragments and the `res-skel.c` files to make `res.c`.

This was because editing a CSV table was easier than editing the C code (it was too easy to loose track of which field was which) and also I thought we could also process the CSV data into a table for the website. (Not sure if that was ever done)

I wouldn't do it that way today, I would probably use JSON or its Lua equivalent to make a tagged data file.

**Comment by rjaguar3** [comment]
Another possibility is to rate possible collections of enemies by difficulty and pick an encounter whose difficulty is at most (your_total_levels + area_difficulty)/2, like NetHack does.  This would allow for encounters to be reasonably tailored to the player characters' ability while still allowing the player to make the game easier by grinding.

I think relying upon hidden caches of money (which, incidentally, I was not even aware of until OnlineCop mentioned them, even though I played the start of the game twice) to balance the first battle is ineffective, as demonstrated by the YouTube videos.

Also, it might be a good idea to switch the current formats for monster and encounter data to a more modern (and flexible) format, like JSON.

**Comment by peterhull90** [comment]
> Another possibility is to rate possible collections of enemies by difficulty and pick an encounter... 
I believe there is something like that; not sure where it is in the code. It may be the assessment of the encounter level is (or has become)  inaccurate.

> Also, it might be a good idea to switch the current formats for monster and encounter data to a more modern (and flexible) format, like JSON.
Agreed. I've used JSON in another project and I liked the flexibility and ease of editing.

**Comment by OnlineCop** [comment]
@peterhull90 Do you recommend a JSON library that we could implement into kq? I've looked at the documentation on some projects like https://github.com/akheron/jansson but didn't see anything specific on how it actually gets implemented into a C-based project itself.

**Comment by peterhull90** [comment]
The one I settled on for the other project was https://github.com/udp/json-parser - it's really simple, just one C file and one header. There are also one implemented in Lua if we chose to go that route. 

On the subject of the original issue, how about dialing down the difficulty of the first couple of encounters and play-testing that? @rjaguar3  would you be able to recompile and test if we provided a patch?

**Comment by rjaguar3** [comment]
@peterhull90 I should be able to test a patch.

**Comment by peterhull90** [comment]
Here is a patch, I took out all the double combats but one; if this is still too much we can reduce that too.

[easier-initial-combat.txt](https://github.com/grrk-bzzt/kq-fork/files/78484/easier-initial-combat.txt)

**Comment by OnlineCop** [comment]
Committed @peterhull90 suggested changes. Commit hash 8bb1512dd4251428b74961bd211e6123bfbb9463.

**Comment by peterhull90** [comment]
I should say this is just an interim change to see how it feels; ultimately it would be good to have a level-based filter as @rjaguar3 suggests.

As far as I can see, in the current code there's a check to avoid combat if it's 'too easy' for the player's current level, and a check to avoid >2 attackers if you are on your own and of insufficient level.

**Comment by peterhull90** [comment]
Just as an aside, if you have a random combat on the main map, KQ

* Gets a zone number (stored in the .map file)
* Translates to a battle number (in the lua script)
* Looks up in the battles table to get an encounter number (hard coded in res.c)
* Looks up in the encounter table to get a monster number (hard coded in res.c)
* Gets monster stats (loaded from allstat.mon and resabil.mon)

I suggest there is scope for simplification here!

**Comment by OnlineCop** [comment]
Create a new ticket/issue ;)

**Comment by z9484** [comment]
Is the obsolete? The first battles seemed okay to me if you scrounge around town to get enough gold to get equipment.

**Comment by OnlineCop** [comment]
I think more balancing is still needed for the first area. The enemies can deal about 10 damage, which is overwhelming with 1 enemy, but almost always deadly when encountering 2 enemies.

These first enemies should be nerfed. Definitely need to drop the attack down to only deal about 3-5 damage apiece, so the player has a chance to level up one to two times before needing any serious healing or an Inn.

Games like Dragon Quest start a player in an area with Slimes and Ants or something, with low HP (usually 3-8) and low attack (1-3 for slimes and about 2-5 for ants), since the player only starts with about 20 HP and has an attack dealing around 2-4 damage apiece at level 1.

**Comment by pedro-w** [comment]
I think it's OK. It is tough but doesn't feel unfair, previously IIRC you could be killed without even getting a combat round yourself. 
And, in any case, once you get in the first tunnel, that's really tough!
